Record subnet saturation in NetworkAttachment status

matte21 opened the pull request:
Make the IPAM controller write into NetworkAttachment status an error
message when the NetworkAttachment's subnet has run out of IP addresses.

matte21 committed:
Record subnet saturation in NetworkAttachment status

Make the IPAM controller write into NetworkAttachment status an error
message when the NetworkAttachment's subnet has run out of IP addresses.

MikeSpreitzer commented on staging/kos/pkg/controllers/ipam/ipam.go:
```diff
@@ -643,7 +670,7 @@ func (ctlr *IPAMController) pickAndLockAddress(ns, name string, att *netv1a1.Net
 	ipForStatusU, ok := ctlr.PickAddress(vni, addrMin, addrMax)
 	if !ok {
-		err = fmt.Errorf("No IP address available in %x/%x--%x for %s/%s", vni, subnetBaseU, subnetLastU, ns, name)
+		err = fmt.Errorf("%s %s (%x/%x--%x)", fullSubnetErrMsgPrefix, subnetName, vni, subnetBaseU, subnetLastU)
```
I think the namespace should be included in this message, and it would not hurt to include the name of the NetworkAttachment.

matte21 replied:
Currently in the IPAM controller code that error message is always embedded in a [message which comprises the namespaced name of the attachment](https://github.com/MikeSpreitzer/kube-examples/blob/write-subnet-saturation-in-attachment-status/staging/kos/pkg/controllers/ipam/ipam.go#L417), so there's no strict need to add the information you're suggesting.
The rule of thumb I try to follow when it comes to error messages is to include no more than what only the producer of the error message knows. If clients of the call which returns the error need an error message with more context they should add it, they are those who know best what's needed. The first error message producer could add all the information right away, but this, in my experience, tends to produce error messages with a lot of duplicated information (higher-level clients will keep embedding the errors they get within other errors).

matte21 replied:
An addition to my previous comment: I agree with adding the subnet namespace. Regardless of my argument on not duplicating info, single messages should be self-contained. But for the argument explained above I think attachment info should not be included.

MikeSpreitzer replied:
Oh, sorry, I had not noticed that this error message gets extended with the NetworkAttachment's namespace and name.  And the subnet's namespace is necessarily the same as the NetworkAttachment's namespace.  So the error message is OK as it is.

I agree with not duplicating.  The way I usually approach that is to make the original producer include all the interesting identification information that it knows, and callers extend with anything they need to add.

In this case that conflicts with the pattern of making the worker loop body add the object namespace and name.

One way to solve this conflict is to make the worker loop not log the errror but rather push that responsibility down the stack.  But that's a bigger change; _if_ we make it then it should maybe go in its own PR.

Or we could just accept duplication in this case for the sake of making really sure that the loop body outcome is logged.

matte21 replied:
> Oh, sorry, I had not noticed that this error message gets extended with the NetworkAttachment's namespace and name. And the subnet's namespace is necessarily the same as the NetworkAttachment's namespace. So the error message is OK as it is.

c2bc019 adds the subnet namespace to the first error message. It's true that it is redundant with the namespace of the network attachment which gets logged in the worker loop body, but it is needed to make the first error message self-contained. Having self-contained error messages  regardless of whether they get printed is a way to future-proof them wrt code refactorings, since the refactored code might print the error message directly without adding the needed context. Of course one has to ponder whether the duplication stemming from such approach is worth it. I usually tend to err on the side of no duplication, but I think in this scenario such a small duplication is ok.

> Or we could just accept duplication in this case for the sake of making really sure that the loop body outcome is logged.

Currently in the referenced log message there's no duplication (with the exception of the aforementioned one), so things are fine as they are. If think the way out of it in general is to never include networkattachments' namespaced names in error messages, knowing that such information will be added in the worker loop. I can't remember what's the current state of KOS controllers in that regard, but I favor a survey and standalone PR to rationalize logging if needed. But I also think this is currently low priority wrt other tasks.

MikeSpreitzer replied:
agreed

MikeSpreitzer commented on staging/kos/pkg/controllers/ipam/ipam.go:
```diff
@@ -465,11 +469,34 @@ func (ctlr *IPAMController) processNetworkAttachment(ns, name string) error {
 	} else {
 		lockForStatus, ipForStatus, err = ctlr.pickAndLockAddress(ns, name, att, subnetName, desiredVNI, desiredBaseU, desiredLastU)
-		if err != nil {
+		if isFullSubnetErr(err) && !fullSubnetMsgFound(att.Status.Errors.IPAM) {
```
This PR is changing subnet-full from something that causes retry to something that determines part of Status.  Regardless of what Status previously contained.  The logic here gets that last part wrong.  We probably want to adopt the pattern of computing what Status should say and then requesting an update if that differs from what Status currently says.

matte21 replied:
> This PR is changing subnet-full from something that causes retry to something that determines part of Status. Regardless of what Status previously contained

Both sentences are false. This PR is changing subnet-full from something that causes retry to something that causes a retry AND determines part of Status. If the subnet is full there will always be a retry, but if the processed NetworkAttachment status already contains an error message describing a full-subnet condition there is no status update.

MikeSpreitzer replied:
Right.

matte21 replied:
Are you convinced of the correctness of the aforementioned code or is there something wrong you still see in it?

MikeSpreitzer replied:
The code is good.  You explained the errors in my original comment.

MikeSpreitzer commented on staging/kos/pkg/controllers/ipam/ipam.go:
```diff
@@ -465,11 +469,34 @@ func (ctlr *IPAMController) processNetworkAttachment(ns, name string) error {
+	err = ctlr.updateNAStatus(ns, name, att, nadat, naStatusErrs, subnetRV, lockForStatus, ipForStatus)
+	if fullSubnetErr != nil {
+		if err != nil {
+			return fmt.Errorf("%s. %s", fullSubnetErr.Error(), err.Error())
```
s/%s. %s/%s; and %s/

Typically the `Error()` of an `error` is an English clause (which is like a sentence but does not have initial capital nor final punctuation and is suitable for composing into a larger clause).

matte21 replied:
I fixed this locally, I'm waiting to push it until we reach consensus on the other issues you brought up to avoid multiple tiny commits in case we chose to change other things as well.

matte21 replied:
Fixed in c2bc019.

MikeSpreitzer reviewed (commented):
A few comments in-line

matte21 committed:
IPAM controller: small fixes

Add subnet namespace to subnet saturation error message for completeness.

Also, do a minor fix to another error message
